File: sdks/ruby/lib/nanoservice/middleware/logging_middleware.rb
# frozen_string_literal: true

module Nanoservice
  module Middleware
    # LoggingMiddleware logs node execution with timing information.
    # It captures start/complete/failed events using the provided Logger.
    class LoggingMiddleware < Middleware
      # @param logger [Nanoservice::Logging::Logger] Logger instance
      def initialize(logger)
        super()
        @logger = logger
      end

      # Wrap the handler with logging behavior.
      # @param handler [#call] Inner handler callable
      # @return [#call] Wrapped callable with logging
      def wrap(handler)
        logger = @logger

        ->(ctx, config) {
          start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
          logger.info("node execution started", fields: { "workflow" => ctx.workflow_name })

          begin
            result = handler.call(ctx, config)
            duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000.0
            logger.info("node execution completed", fields: {
              "workflow"    => ctx.workflow_name,
              "duration_ms" => duration_ms.round(2)
            })
            result
          rescue StandardError => e
            duration_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000.0
            logger.error("node execution failed", fields: {
              "workflow"    => ctx.workflow_name,
              "duration_ms" => duration_ms.round(2),
              "error"       => e.message
            })
            raise
          end
        }
      end
    end
  end
end

File: sdks/csharp/src/Nanoservice.Core/Middleware/LoggingMiddleware.cs
using System.Diagnostics;
using System.Text.Json;
using Nanoservice.Core.Node;

namespace Nanoservice.Core.Middleware;

/// <summary>
/// LoggingMiddleware logs node execution with timing.
/// </summary>
public class LoggingMiddleware : IMiddleware
{
    public INodeHandler Wrap(INodeHandler next)
    {
        return new LoggingHandler(next);
    }

    private class LoggingHandler : INodeHandler
    {
        private readonly INodeHandler _inner;

        public LoggingHandler(INodeHandler inner)
        {
            _inner = inner;
        }

        public async Task<JsonElement> ExecuteAsync(Types.Context ctx, Dictionary<string, JsonElement> config)
        {
            var stopwatch = Stopwatch.StartNew();
            Console.WriteLine($"[INFO] {DateTime.UtcNow:O} Node execution started workflow={ctx.WorkflowName}");

            try
            {
                var result = await _inner.ExecuteAsync(ctx, config);
                stopwatch.Stop();
                Console.WriteLine($"[INFO] {DateTime.UtcNow:O} Node execution completed workflow={ctx.WorkflowName} duration_ms={stopwatch.Elapsed.TotalMilliseconds:F2}");
                return result;
            }
            catch (Exception ex)
            {
                stopwatch.Stop();
                Console.WriteLine($"[ERROR] {DateTime.UtcNow:O} Node execution failed workflow={ctx.WorkflowName} duration_ms={stopwatch.Elapsed.TotalMilliseconds:F2} error={ex.Message}");
                throw;
            }
        }
    }
}
